undefined reference to GetMessageA when using mingw64 and cmake

<!--

NOTE: this form is for bug reports only. For questions or troubleshooting, please see the protobuf mailing list: https://groups.google.com/forum/#!forum/protobuf

-->

**What version of protobuf and what language are you using?**
Version: master/v3.17.3
Language: C++

**What operating system (Linux, Windows, ...) and version?**
Window 10

**What runtime / compiler are you using (e.g., python version or gcc version)**
gcc.exe (GCC with MCF thread model, built by LH_Mouse.) 11.1.1 20210604

**What did you do?**
Steps to reproduce the behavior:
1. open protobuf/cmake using vscode
2. build with vscode cmake tools

we got the following errors while linking tests.ext

`[build] [ 99%] Linking CXX executable tests.exe
[build] C:/Programs/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.1.1/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\tests.dir/objects.a(map_test.cc.obj):D:/Workspace/C_CPP/protobuf-3.17.3/src/google/protobuf/map_test.cc:1315: undefined reference to `google::protobuf::Reflection::GetMessageA(google::protobuf::Message const&, google::protobuf::FieldDescriptor const*, google::protobuf::MessageFactory*) const'
[build] C:/Programs/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.1.1/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\tests.dir/objects.a(map_test.cc.obj): in function `TestBody':
[build] D:/Workspace/C_CPP/protobuf-3.17.3/src/google/protobuf/map_test.cc:1351: undefined reference to `google::protobuf::Reflection::GetMessageA(google::protobuf::Message const&, google::protobuf::FieldDescriptor const*, google::protobuf::MessageFactory*) const'
[build] C:/Programs/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.1.1/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\tests.dir/objects.a(map_test.cc.obj):D:/Workspace/C_CPP/protobuf-3.17.3/src/google/protobuf/map_test.cc:1542: undefined reference to `google::protobuf::Reflection::GetMessageA(google::protobuf::Message const&, google::protobuf::FieldDescriptor const*, google::protobuf::MessageFactory*) const'
[build] C:/Programs/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.1.1/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\tests.dir/objects.a(map_test.cc.obj): in function `TestBody':
[build] D:/Workspace/C_CPP/protobuf-3.17.3/src/google/protobuf/map_test.cc:1582: undefined reference to `google::protobuf::Reflection::GetMessageA(google::protobuf::Message const&, google::protobuf::FieldDescriptor const*, google::protobuf::MessageFactory*) const'
[build] C:/Programs/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.1.1/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/Workspace/C_CPP/protobuf-3.17.3/src/google/protobuf/map_test.cc:1710: undefined reference to `google::protobuf::Reflection::GetMessageA(google::protobuf::Message const&, google::protobuf::FieldDescriptor const*, google::protobuf::MessageFactory*) const'
[build] C:/Programs/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.1.1/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\tests.dir/objects.a(map_test.cc.obj):D:/Workspace/C_CPP/protobuf-3.17.3/src/google/protobuf/map_test.cc:1865: more undefined references to `google::protobuf::Reflection::GetMessageA(google::protobuf::Message const&, google::protobuf::FieldDescriptor const*, google::protobuf::MessageFactory*) const' follow
[build] collect2.exe: error: ld returned 1 exit status
[build] mingw32-make[2]: *** [CMakeFiles\tests.dir\build.make:2401: tests.exe] Error 1
[build] mingw32-make[1]: *** [CMakeFiles\Makefile2:262: CMakeFiles/tests.dir/all] Error 2
[build] mingw32-make: *** [Makefile:135: all] Error 2
[build] Build finished with exit code 2
`

--- a/src/google/protobuf/map_test.cc
+++ b/src/google/protobuf/map_test.cc
@@ -30,7 +30,7 @@
 
 // A hack to include windows.h first, which ensures the GetMessage macro can
 // be undefined when we include <google/protobuf/stubs/common.h>
-#if defined(_WIN32)
+#if defined(_MSC_VER)
 #define _WINSOCKAPI_  // to avoid re-definition in WinSock2.h
 #define NOMINMAX      // to avoid defining min/max macros
 #include <windows.h>

